### PR TITLE
feat: added support for redis cluster (https://www.npmjs.com/package/redis)

### PR DESCRIPTION
**Pull request by kirrg001**
refs https://jsw.ibm.com/browse/INSTA-9475

- [x] fix connection issue -> reply https://github.com/redis/node-redis/issues/2815
- [x] ioredis (separate PR)
- [x] CI sidecars (https://github.com/bitnami/charts/issues/28894) - use cloud cluster for now
- [x] redis vs @redis/client

# Summary
We use the cluster from Azure for localhost and Tekton.
See https://github.com/bitnami/charts/issues/28894
See https://github.com/redis/node-redis/issues/2815

We keep the docker redis cluster config. Does not hurt to keep it.

Figured out, you can use the redis pkg either with `require('redis')` or with `@redis/client`. Added support for this too.
I will merge both features as separate commits.

Cluster support for `ioredis` coming in a different PR.

### CODE REVIEW

File: packages/collector/test/tracing/database/redis/connect-via/cluster.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+// node bin/start-test-containers.js --redis-node-0 --redis-node-1 --redis-node-2
+// docker exec -it 2aaaac7b9112 redis-cli -p 6379 cluster info
+module.exports = async function connect(redis, log, tries = 0) {
+  // NOTE: we cannot run redis cluster on Tekton https://github.com/bitnami/charts/issues/28894
+  // NOTE: we cannot run redis cluster on Docker locally https://github.com/redis/node-redis/issues/2815
+  //
+  // Please set the environment variables to connect to the cloud redis cluster:
+  //    export AZURE_REDIS_CLUSTER=team-nodejs-redis-cluster-tekton.redis.cache.windows.net:6380
+  //    export AZURE_REDIS_CLUSTER_PWD=
```
Comment (aryamohanan) [inline packages/collector/test/tracing/database/redis/connect-via/cluster.js:18]: Are these credentials available on 1Password? It would be helpful if you could mention that as well.

File: packages/core/test/util/initializedTooLateHeuristic_test.js [hunk 2]
```diff
@@ -54,7 +54,7 @@ describe('[UNIT] util.initializedTooLateHeurstic', function () {
   it('hasBeenInitializedTooLate is true', () => {
     instrumentedModules.forEach(moduleName => {
       // eslint-disable-next-line no-console
-      console.log(`for module ${moduleName}`);
+      console.log(`### for module ${moduleName}`);
```
Comment (aryamohanan) [inline packages/core/test/util/initializedTooLateHeuristic_test.js:57]: Was this change intentional?

File: packages/collector/test/tracing/database/redis/connect-via/cluster.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+'use strict';
+
+const MAX_TRIES = 50;
+const { delay } = require('../../../../../../core/test/test_util');
+
+// node bin/start-test-containers.js --redis-node-0 --redis-node-1 --redis-node-2
+// docker exec -it 2aaaac7b9112 redis-cli -p 6379 cluster info
+module.exports = async function connect(redis, log, tries = 0) {
+  // NOTE: we cannot run redis cluster on Tekton https://github.com/bitnami/charts/issues/28894
```
Comment (aryamohanan) [inline packages/collector/test/tracing/database/redis/connect-via/cluster.js:13]: The comments are little confusing, how can we run the tests in local?

File: packages/collector/test/tracing/database/redis/connect-via/cluster.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
+/*
+ * (c) Copyright IBM Corp. 2024
+ */
+
+'use strict';
+
+const MAX_TRIES = 50;
+const { delay } = require('../../../../../../core/test/test_util');
+
+// node bin/start-test-containers.js --redis-node-0 --redis-node-1 --redis-node-2
```
Comment (aryamohanan) [inline packages/collector/test/tracing/database/redis/connect-via/cluster.js:10]: Does this command work locally?

File: packages/collector/test/tracing/database/redis/connect-via/default.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+/*
+ * (c) Copyright IBM Corp. 2024
+ */
+
+'use strict';
+
+module.exports = async function (redis, log) {
+  const client = redis.createClient({ url: `redis://${process.env.REDIS}` });
```
Comment (aryamohanan) [inline packages/collector/test/tracing/database/redis/connect-via/default.js:8]: It would be better to add comments on how to run the redis tests locally.

**Review by aryamohanan — CHANGES_REQUESTED** [review]

File: packages/collector/test/tracing/database/redis/connect-via/cluster.js (hunk 1 above)
Comment (kirrg001) [reply packages/collector/test/tracing/database/redis/connect-via/cluster.js:18]: > Are these credentials available on 1Password? 

Yes sure!

File: packages/core/test/util/initializedTooLateHeuristic_test.js (hunk 2 above)
Comment (kirrg001) [reply packages/core/test/util/initializedTooLateHeuristic_test.js:57]: Yes. If you run the test, its hard to understand the outputs.

File: packages/collector/test/tracing/database/redis/connect-via/cluster.js (hunk 4 above)
Comment (kirrg001) [reply packages/collector/test/tracing/database/redis/connect-via/cluster.js:10]: Yes

File: packages/collector/test/tracing/database/redis/connect-via/cluster.js (hunk 3 above)
Comment (kirrg001) [reply packages/collector/test/tracing/database/redis/connect-via/cluster.js:13]: > We use the cluster from Azure for localhost and Tekton.

https://github.com/instana/nodejs/pull/1270/files#diff-0ded018ffc0132683fbe045fceb5018de0890e0b2a318bb2bcb801ddaffca85cR16

I will update the comment section to make it more clear 👍 
The docker command is the initial command I have used.

File: packages/collector/test/tracing/database/redis/connect-via/default.js (hunk 5 above)
Comment (kirrg001) [reply packages/collector/test/tracing/database/redis/connect-via/default.js:8]: Already discussed in previous comment.

File: packages/collector/test/tracing/database/redis/connect-via/cluster.js (hunk 3 above)
Comment (kirrg001) [reply packages/collector/test/tracing/database/redis/connect-via/cluster.js:13]: >  // Please set the environment variables to connect to the cloud redis cluster:
>  //    export AZURE_REDIS_CLUSTER=team-nodejs-redis-cluster-tekton.redis.cache.windows.net:6380
>  //    export AZURE_REDIS_CLUSTER_PWD=

File: packages/collector/test/tracing/database/redis/test.js [hunk 6]
```diff
@@ -21,748 +22,783 @@ const {
...
+                    qs: {
+                      key: 'price'
+                    }
+                  })
+                )
+                .then(response => {
+                  expect(String(response)).to.equal('42');
+
+                  return retry(() =>
+                    agentControls.getSpans().then(spans => {
```
Comment (aryamohanan) [inline packages/collector/test/tracing/database/redis/test.js:107]: Question: Should we consider adding a check for the count of the spans?

**Review by aryamohanan — APPROVED** [review]

File: packages/collector/test/tracing/database/redis/test.js (hunk 6 above)
Comment (kirrg001) [reply packages/collector/test/tracing/database/redis/test.js:107]: Yeah lets add that 👍 Good catch
